presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1411,48 +1413,78 @@ else if (expression instanceof LongLiteral) {
         return orderByExpressions;
     }
 
-    private List<FieldOrExpression> analyzeGroupBy(QuerySpecification node, RelationType tupleDescriptor, AnalysisContext context, List<FieldOrExpression> outputExpressions)
+    private List<List<FieldOrExpression>> analyzeGroupBy(QuerySpecification node, RelationType tupleDescriptor, AnalysisContext context, List<FieldOrExpression> outputExpressions)
     {
-        ImmutableList.Builder<FieldOrExpression> groupByExpressionsBuilder = ImmutableList.builder();
-        if (!node.getGroupBy().isEmpty()) {
-            // Translate group by expressions that reference ordinals
-            for (Expression expression : node.getGroupBy()) {
-                // first, see if this is an ordinal
-                FieldOrExpression groupByExpression;
+        List<Set<Set<Expression>>> enumeratedGroupingSets = node.getGroupBy().stream()
+                .map(groupingElement -> groupingElement.enumerateGroupingSets().stream()
+                        .collect(Collectors.<Set<Expression>>toSet()))
+                .distinct()
+                .collect(toImmutableList());
 
-                if (expression instanceof LongLiteral) {
-                    long ordinal = ((LongLiteral) expression).getValue();
-                    if (ordinal < 1 || ordinal > outputExpressions.size()) {
-                        throw new SemanticException(INVALID_ORDINAL, expression, "GROUP BY position %s is not in select list", ordinal);
-                    }
+        // compute cross product of enumerated grouping sets
+        List<List<Expression>> computedGroupingSets = Sets.cartesianProduct(enumeratedGroupingSets).stream()
+                .map(groupingSetList -> groupingSetList.stream()
+                        .flatMap(Collection::stream)
+                        .distinct()
+                        .collect(toImmutableList()))
+                .distinct()
+                .collect(toImmutableList());
 
-                    groupByExpression = outputExpressions.get((int) (ordinal - 1));
-                }
-                else {
-                    ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, tupleDescriptor, context);
-                    analysis.addInPredicates(node, expressionAnalysis.getSubqueryInPredicates());
-                    groupByExpression = new FieldOrExpression(expression);
-                }
+        // check if there are any columns at all to group by
+        if (computedGroupingSets.size() == 1 && computedGroupingSets.get(0).isEmpty()) {
+            List<List<FieldOrExpression>> analyzedGroupingSets = extractAggregates(node).isEmpty() ? ImmutableList.of() : ImmutableList.of(ImmutableList.of());
```

Comment:
Can you explain what this line does? I'm not sure I follow.

If there is at least one effective grouping set, this is an aggregation query, so there's no need to check for aggregates. If there are no effective grouping sets (i.e., because of a missing group by clause), *then* we need to check if there are aggregates and create a synthetic `()` grouping set